processing/src/main/java/org/apache/druid/query/DefaultQueryMetrics.java

```diff
@@ -24,50 +24,89 @@
-  protected final Map<String, Number> metrics = new HashMap<>();
+  private static final EmittingLogger log = new EmittingLogger(DefaultQueryMetrics.class);
 
-  /** Non final to give subclasses ability to reassign it. */
-  protected Thread ownerThread = Thread.currentThread();
+  protected final ObjectMapper jsonMapper;
+  protected final Object lock = new Object();
+  @GuardedBy("lock") private final Map<String, String> singleValueDims = new HashMap<>();
+  @GuardedBy("lock") private final Map<String, String[]> multiValueDims = new HashMap<>();
+  @GuardedBy("lock") private final Map<String, Number> metrics = new HashMap<>();
 
   public DefaultQueryMetrics(ObjectMapper jsonMapper)
   {
     this.jsonMapper = jsonMapper;
   }
 
-  protected void checkModifiedFromOwnerThread()
+  // copy constructor, used by makeCopy()
+  public DefaultQueryMetrics(DefaultQueryMetrics that)
   {
-    if (Thread.currentThread() != ownerThread) {
-      throw new IllegalStateException(
-          "DefaultQueryMetrics must not be modified from multiple threads. If it is needed to gather dimension or "
-          + "metric information from multiple threads or from an async thread, this information should explicitly be "
-          + "passed between threads (e. g. using Futures), or this DefaultQueryMetrics's ownerThread should be "
-          + "reassigned explicitly");
-    }
+    this.jsonMapper = that.jsonMapper;
+    this.singleValueDims.putAll(that.singleValueDims);
+    this.multiValueDims.putAll(that.multiValueDims);
+    this.metrics.putAll(that.metrics);
   }
 
   protected void setDimension(String dimension, String value)
   {
-    checkModifiedFromOwnerThread();
-    builder.setDimension(dimension, value);
+    synchronized (lock) {
+      String oldValue = singleValueDims.put(dimension, value);
+      if (oldValue != null && !oldValue.equals(value)) {
+        handleIllegalModification(singleValueDims.getOrDefault(DruidMetrics.ID, ""), dimension);
+      }
+    }
+  }
+
+  protected void setDimensions(String dimension, String[] values)
+  {
+    synchronized (lock) {
+      String[] oldValues = multiValueDims.put(dimension, values);
+      if (oldValues != null && !Arrays.equals(oldValues, values)) {
+        handleIllegalModification(singleValueDims.getOrDefault(DruidMetrics.ID, ""), dimension);
+      }
+    }
+  }
+
+  protected QueryMetrics<QueryType> reportMetric(String metricName, Number value)
+  {
+    synchronized (lock) {
+      Number oldValue = metrics.put(metricName, value);
+      if (oldValue != null && !oldValue.equals(value)) {
+        handleIllegalModification(singleValueDims.getOrDefault(DruidMetrics.ID, ""), metricName);
+      }
+      return this;
+    }
+  }
+
+  private void handleIllegalModification(String queryId, String entryName)
+  {
+    Exception e = new Exception("stack trace");
+    log.makeAlert(e, "\"%s\" in QueryMetrics got modified to another value", entryName)
```

Comment:
Should throw an `IllegalStateException`, not just log it. Errors of this kind should remain very verbose.

Reply:
I didn't throw exception here because I think bugs in metrics handling shouldn't fail the whole query. It would allow users continue to query with only "broken" metrics, which may be acceptable in some use cases. I admit that this may be controversial and would like to hear other committer's opinions.

Reply:
OK, will throw IllegalStateException instead